### PR TITLE
[#3580] ipasswd issues (was: stty path hard-coded in clientLogin.cpp)

### PR DESCRIPTION
**Pull request by andrew-irods**
CI test successful: http://172.25.14.125:8080/view/2.%20Personal/job/irods-build-and-test-workflow/861/

### CODE REVIEW

File: lib/core/src/obf.cpp [hunk 1]
```diff
@@ -289,30 +292,34 @@ obfSavePw( int promptOpt, int fileOpt, int printOpt, const char *pwArg ) {
 
     envVal = obfiGetEnvKey();
 
-    if ( strlen( pwArg ) == 0 ) {
-
+    if ( pwArg == NULL || strlen( pwArg ) == 0 ) {
```
Comment (trel) [inline lib/core/src/obf.cpp]: more Yoda?

File: lib/core/src/obf.cpp [hunk 2]
```diff
@@ -289,30 +292,34 @@ obfSavePw( int promptOpt, int fileOpt, int printOpt, const char *pwArg ) {
...
+            {
+                printf( "WARNING: Error %d disabling echo mode. Password will be displayed in plaintext.\n", tiosutl.getError() );
             }
         }
 
         printf( "Enter your current iRODS password:" );
-        if ( NULL == fgets( inbuf, MAX_PASSWORD_LEN + 50, stdin ) ) {
-            // end of line reached or no input
-        }
+        const char *fgets_return = fgets( inbuf, MAX_PASSWORD_LEN + 50, stdin );
```
Comment (trel) [inline lib/core/src/obf.cpp]: magic number?

File: lib/core/include/termiosUtil.hpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                else
+                {
+                    errno_copy_ = errno;
+                    valid_ = false;
+                }
+            };
+
+        private:
+            // Invalid constructor.  Ensures instantiation with fd parameter
+            termiosUtil(void) {};
```
Comment (korydraughn) [inline lib/core/include/termiosUtil.hpp]: Your constructor accepting a file descriptor will stop the compiler from generating a default constructor, so the private default constructor can be removed.

File: lib/core/include/termiosUtil.hpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        public:
+            virtual ~termiosUtil(void) = default;
```
Comment (korydraughn) [inline lib/core/include/termiosUtil.hpp]: Is it necessary to declare the destructor as virtual?  Looking over this pull request, I haven't spotted a need for this.  Are you expecting this class to be used as a base?

File: lib/core/include/termiosUtil.hpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                errno_copy_ = errno;
+                valid_ = false;
+                return false;
+            }
+
+       private:
+            bool valid_;            // fd_ >= 0, etc
+            int errno_copy_;        // if fd_ is not a tty, or other termios error, this has the error
+            int fd_;                // File descriptor for this tty_ structure
+            struct termios tty_;
```
Comment (korydraughn) [inline lib/core/include/termiosUtil.hpp]: Is `struct` needed here?

File: lib/core/include/termiosUtil.hpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        private:
+            // Invalid constructor.  Ensures instantiation with fd parameter
+            termiosUtil(void) {};
+
+        public:
+            virtual ~termiosUtil(void) = default;
+            int getError(void) const     { return errno_copy_; };
+            bool getValid(void) const     { return valid_; };
+
+            bool echoOff(void)
```
Comment (korydraughn) [inline lib/core/include/termiosUtil.hpp]: `void` is probably not needed.  An empty set of parenthesis is pretty clear.

File: lib/core/src/obf.cpp (hunk 1 above)
Comment (andrew-irods) [reply lib/core/src/obf.cpp]: Discuss this we did. Change this I will not.

File: lib/core/include/termiosUtil.hpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
+#ifndef _TERMIOS_UTIL_HPP_
+#define _TERMIOS_UTIL_HPP_
+
+#include <unistd.h>
+#include <errno.h>
+#include <termios.h>
+
+#ifdef __cplusplus
```
Comment (korydraughn) [inline lib/core/include/termiosUtil.hpp]: Why is this needed?

File: lib/core/src/obf.cpp (hunk 2 above)
Comment (andrew-irods) [reply lib/core/src/obf.cpp]: "Not my fault".  I just added the examination of the return value.  But I know what you mean.

File: lib/core/include/termiosUtil.hpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace irods
+{
+    // class termiosUtil - at this time, this class encapsulates the work
+    // which has to be done in order to turn echo on or off for a specifc
+    // file desctiptor.  The call can be expanded to encapsulate other
```
[the remark this answers is not in the document]
Comment (andrew-irods) [reply lib/core/include/termiosUtil.hpp]: Will change.  Thanks.

File: lib/core/include/termiosUtil.hpp (hunk 3 above)
Comment (andrew-irods) [reply lib/core/include/termiosUtil.hpp]: Will change.  Thanks.

File: lib/core/include/termiosUtil.hpp (hunk 4 above)
Comment (andrew-irods) [reply lib/core/include/termiosUtil.hpp]: Yes yes and yes.  I don't know that it won't be used as a base class 2 years from now.  (My opinion, which means that this stays).

File: lib/core/include/termiosUtil.hpp (hunk 5 above)
Comment (andrew-irods) [reply lib/core/include/termiosUtil.hpp]: No it's not.  It's staying (it's a choice, not a rationale).

File: lib/core/include/termiosUtil.hpp (hunk 6 above)
Comment (andrew-irods) [reply lib/core/include/termiosUtil.hpp]: Again - this is a choice.  It stays.

File: lib/core/include/termiosUtil.hpp (hunk 7 above)
Comment (andrew-irods) [reply lib/core/include/termiosUtil.hpp]: Over abundance of being careful.

File: lib/core/include/termiosUtil.hpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                {
+                    errno_copy_ = errno;
+                    valid_ = false;
+                }
+            };
+
+        public:
+            virtual ~termiosUtil(void) = default;
+            int getError(void) const     { return errno_copy_; };
+            bool getValid(void) const     { return valid_; };
```
Comment (korydraughn) [inline lib/core/include/termiosUtil.hpp]: After reviewing the pull request for the icommand [ipasswd], I think `getValid()` should be changed to `valid()` or `isValid()`.
